src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -2031,7 +2032,30 @@ WorkerExtendedOpNode(MultiExtendedOp *originalOpNode,
 	workerExtendedOpNode->hasWindowFuncs = originalOpNode->hasWindowFuncs;
 	workerExtendedOpNode->windowClause = originalOpNode->windowClause;
 
-	if (!queryHasAggregates)
+	/*
+	 * we are only pushing down distinct on, and limit (along with order by)
+	 * if it distinct on clause is on bare column. If it has expression, aggregates
+	 * or function calls, they are evaluated at worker query and returned to master
+	 * to perform necessary distinct on/order by/limit operations.
+	 */
+	if (originalOpNode->hasDistinctOn)
```

Comment:
Looks like we've wrong results for queries without an expression on the `DISTINCT`
```SQL
-- wrong results without expression in the DISTINCT
SELECT DISTINCT ON (value_2) value_2, user_id, avg(value_1)
FROM 
	users_table
GROUP BY 
	user_id, value_1, value_2
ORDER BY 
	1 DESC, 3 DESC, 2 DESC
LIMIT 10;
```